packages/react-test-renderer/src/ReactTestRenderer.js

```diff
@@ -564,40 +560,48 @@ const ReactTestRendererFiber = {
 
   unstable_setNowImplementation: setNowImplementation,
 
-  act(callback: () => void): Thenable {
+  act(callback: () => void | Promise<void>) {
     // note: keep these warning messages in sync with
-    // createNoop.js and ReactTestUtils.js
-    let result = batchedUpdates(callback);
-    if (__DEV__) {
-      if (result !== undefined) {
-        let addendum;
-        if (result !== null && typeof result.then === 'function') {
-          addendum =
-            "\n\nIt looks like you wrote TestRenderer.act(async () => ...) or returned a Promise from it's callback. " +
-            'Putting asynchronous logic inside TestRenderer.act(...) is not supported.\n';
-        } else {
-          addendum = ' You returned: ' + result;
-        }
-        warningWithoutStack(
-          false,
-          'The callback passed to TestRenderer.act(...) function must not return anything.%s',
-          addendum,
-        );
+    // createReactNoop.js and ReactTestUtils.js
+    const result = actedUpdates(callback);
+    if (result && result.then) {
```

Comment:
We already have a definition for how to test a thenable:
https://github.com/facebook/react/blob/master/packages/react-reconciler/src/ReactFiberUnwindWork.js#L203-L205